---
title: _signatureSearch_ package
author: "Author: Yuzhu Duan (yduan004@ucr.edu)"
date: "Last update: `r format(Sys.time(), '%d %B, %Y')`" 
output:
  pdf_document:
    toc: true
  html_document:
    toc: true
    toc_float:
        collapsed: false
        smooth_scroll: true
    toc_depth: 4
    fig_caption: yes
    code_folding: show
    number_sections: true
fontsize: 15pt
bibliography: bibtex.bib
---

<style>
pre code, pre, code {
  white-space: pre !important;
  overflow-x: auto !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>

```{r setup0, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(width=200)
```

```{r setup, echo=FALSE, messages=FALSE, warnings=FALSE}
suppressPackageStartupMessages({
  library(signatureSearch)
  library(HDF5Array)
})
```

# Background

This project is about optimizing signature search and enrichment methods for the discovery of novel modes of action (MOA) of bioactive compounds from reference databases, such as LINCS, containing the genome-wide gene expression signatures (GESs) from tens of thousands of drug and genetic perturbations. The methods used by this prediction workflow can be divided into two major classes. First, gene expression signature search (GESS) methods are used to identify drugs that induce GESs similar to those of query GESs of interest. The queries can be drug- or disease-related GESs. Since the MOA of most drugs in the corresponding reference databases are known, the resulting associations are useful to gain insights into pharmacological and disease mechanisms and to develop novel drug repurposing approaches. Second, functional enrichment analysis (FEA) methods using Gene Ontology (GO) or pathway annotations have been developed to functionally interpret the vast number of GESS results generated by this project. The latter are composed of lists of drugs ranked by the similarity metric of the corresponding GESS method making the functional interpretation of their top ranking drugs challenging. Importantly, the FEA methods developed by this study also support the reconstruction of drug-target networks to guide the interpretation of the results.

## Gene Expression Signature Databases

\cite{Lamb2006-du} generated a GES database, initially including 164 drugs screened against four mammalian cell lines \citep{Lamb2006-du}. A few years later it was extended to 1,309 drugs and eight cell lines. In 2017, the database was increased to 19,811 drugs by the Library of Network-Based Cellular Signatures (LINCS) Consortium \citep{Subramanian2017-fu}, while the number of cell types represented in the database has increased to over 70 normal and cancer cell lines. The number of compound dosages and time points considered in the assays has also been increased by 10-20 fold. The initial database used Affymetrix Gene Chips as expression platform. To scale from a few thousand to many hundred thousand GESs, the LINCS Consortium uses now the more economic L1000 assay. This bead-based technology is a low cost, high-throughput reduced representation expression profiling assay. It measures the expression of 978 landmark genes and 80 control genes by detecting fluorescent intensity of beads after capturing the ligation-mediated amplification products of mRNAs \citep{Peck2006-rf}. The expression of 11,350 additional genes is imputed from the landmark genes by using as training data a collection of 12,063 Affymetrix gene chips \citep{Edgar2002-di}. In the \cite{Subramanian2017-fu} paper the authors use a subset of the LINCS database, referred to as Touchstone, containing 2,837 drug treatments of nine cell lines that can be searched on their project's website (CLUE). The substantial scale-up of the LINCS project provides now many new opportunities to explore MOAs for a vast number of small molecules.

# Terminology

Gene Expression Signatures (GESs): can be Gene Expression Profiles (GEPs), which can be genome-wide profile from differential expression analysis (log2FC or Z-scores) or gene expression intensity values from perturbagen treatment in cell culture, or Differentially Expressed Gene sets (DEGs) from a treatment.

# Database

The CMAP database is stored as \code{SummarizedExperiment} object, which is HDF5 backed on disk and can be loaded via `loadHDF5SummarizedExperiment` function in `rhdf5` package.  
##' For \code{SummarizedExperiment} object, The 'assays' slot should store a \code{DelayedArray} object consist of genome-wide gene expression 
##' profiles (GEPs) from various drug treatments or genetic perturbations. 
##' It represents reference database that the query signature is searched against. 
##' Can be existing public databases (CMAP or LINCS) or custom database.
##' 
##' `CMAP` HDF5 backend database can be downloaded at \url{http://biocluster.ucr.edu/~yduan004/CMAP_db/cmap.tar.gz}, 
#' untar the file by "tar -xzvf file.tar.gz" command, then load the \code{SummarizedExperiment} object via 
##' `loadHDF5SummarizedExperiment` function. The 'assays' slot of the loaded SummarizedExperiment object contains the logFC scores generated 
##' from differential expression (DE) analysis with `limma` of the original CEL files generated by Affymetrix Chips. 
##' The CMAP Affy data is stored at \url{https://portals.broadinstitute.org/cmap/}.
#' The code used for DE analysis is vailable at \url{http://girke.bioinformatics.ucr.edu/longevityTools/mydoc/mydoc_longevityTools_eDRUG_05.html}
#' 
#' `LINCS` HDF5 backend database can be downloaded at \url{http://biocluster.ucr.edu/~yduan004/LINCS_db/lincs42.tar.gz},
#' untar the file by "tar -xzvf file.tar.gz" command, then load the \code{SummarizedExperiment} object via 
##' `loadHDF5SummarizedExperiment` function.

# Methods for GESS

## Get longevity related query signature from Peters et al. (2015)

The query signature are 150 up and 150 down regulated gene sets from 1,497 age-related genes after differential expression analysis
```{r peter_sig, eval=FALSE}
## Obtain 150 up and 150 down regulated gene sets from Peters
Peters <- read.delim("~/insync/project/longevityTools_eDRUG/data/PMID26490707_S1.xls", comment="#", check.names=FALSE)    
#Peters <- read.delim("/rhome/tgirke/Projects/longevity/longevityTools/vignettes/data/PMID26490707_S1.xls", comment="#", check.names=FALSE)
Peters <- Peters[!duplicated(Peters[,"NEW-Entrez-ID"]),]         
query <- Peters$Zscore; names(query) <- Peters[,"NEW-Entrez-ID"]
upset <- head(names(query[order(-query)]), 150)
downset <- tail(names(query[order(-query)]), 150)

# Load public CMAP, LINCS and LINCS_sub database
se_cmap <- loadHDF5SummarizedExperiment("~/insync/project/lincs_gse92742_dataset_analysis/data/cmap")
se_lincs <- loadHDF5SummarizedExperiment("~/insync/project/lincs_gse92742_dataset_analysis/data/lincs42")
se_lincs_sub <- loadHDF5SummarizedExperiment("~/insync/project/lincs_gse92742_dataset_analysis/data/lincs42_sub")

# Load public CMAP_expr, LINCS_expr and LINCS_sub_expr database
se_cmap_expr <- loadHDF5SummarizedExperiment("~/insync/project/lincs_gse92742_dataset_analysis/data/cmap_expr")
se_lincs_expr <- loadHDF5SummarizedExperiment("~/insync/project/lincs_gse92742_dataset_analysis/data/lincs42_expr")
se_lincs_sub_expr <- loadHDF5SummarizedExperiment("~/insync/project/lincs_gse92742_dataset_analysis/data/lincs42_sub_expr")
```

View cell type information in LINCS database
```{r cell_info, eval=FALSE}
cell_info <- metadata(se_lincs)$cell_info
cell_info
```
## Searching with CMAP method for GESS

Generate "qSig" object for GESS_CMAP method and search against CMAP, LINCS, LINCS_sub reference database
```{r gess_cmap, eval=FALSE}
# Search against CMAP reference database
qsig_cmap_against_cmap <- qSig(qsig = list(upset=upset, downset=downset), gess_method = "CMAP", refdb = se_cmap)
cmap_against_cmap <- gess_cmap(qSig=qsig_cmap_against_cmap, chunk_size=5000)

# Search against LINCS reference database
qsig_cmap_against_lincs <- qSig(qsig = list(upset=upset, downset=downset), gess_method = "CMAP", refdb = se_lincs)
cmap_against_lincs <- gess_cmap(qSig=qsig_cmap_against_lincs, chunk_size=5000)

# Search against LINCS_sub reference database
qsig_cmap_against_lincs_sub <- qSig(qsig = list(upset=upset, downset=downset), gess_method = "CMAP", refdb = se_lincs_sub)
cmap_against_lincs_sub <- gess_cmap(qSig=qsig_cmap_against_lincs_sub, chunk_size=5000)
```

```{r show_cmap_res, eval=FALSE}
cmap_against_cmap <- readRDS("~/insync/project/GESS_and_FEA/results/cmap_against_cmap.rds")
cmap_against_cmap@result
cmap_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/cmap_against_lincs.rds")
cmap_against_lincs@result
cmap_against_lincs_sub <- readRDS("~/insync/project/GESS_and_FEA/results/cmap_against_lincs_sub.rds")
cmap_against_lincs_sub@result
```

## Searching with LINCS method for GESS

Generate "qSig" object for GESS_LINCS method and search against CMAP, LINCS, LINCS_sub reference database
```{r gess_lincs, eval=FALSE}
# Search against CMAP reference database
qsig_lincs_against_cmap <- qSig(qsig = list(upset=upset, downset=downset), gess_method = "LINCS", refdb = se_cmap)
lincs_against_cmap <- gess_lincs(qsig_lincs_against_cmap, sortby="WTCS")

# Search against LINCS reference database
qsig_lincs_against_lincs <- qSig(qsig = list(upset=upset, downset=downset), gess_method = "LINCS", refdb = se_lincs)
lincs_against_lincs <- gess_lincs(qSig=qsig_lincs_against_lincs, ES_NULL="Default", taurefList="Default", sortby="NCS", chunk_size=5000)

# Search against LINCS_sub reference database
qsig_lincs_against_lincs_sub <- qSig(qsig = list(upset=upset, downset=downset), gess_method = "LINCS", refdb = se_lincs_sub)
lincs_against_lincs_sub <- gess_lincs(qsig_lincs_against_lincs_sub, sortby="NCS")
```

```{r show_lincs_res, eval=FALSE}
lincs_against_cmap <- readRDS("~/insync/project/GESS_and_FEA/results/lincs_against_cmap.rds")
lincs_against_cmap@result
lincs_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/lincs_against_lincs.rds")
lincs_against_lincs@result
lincs_against_lincs_sub <- readRDS("~/insync/project/GESS_and_FEA/results/lincs_against_lincs_sub.rds")
lincs_against_lincs_sub@result
```

## Searching with gCMAP method for GESS

Generate "qSig" object for GESS_gCMAP method and search against CMAP, LINCS, LINCS_sub reference database
```{r gess_gcmap, eval=FALSE}
query <- as.matrix(assay(se_lincs)[,"sirolimus__MCF7__trt_cp"])

# Search against CMAP reference database
qsig_gcmap_against_cmap <- qSig(qsig = as.matrix(query), gess_method = "gCMAP", refdb = se_cmap)
gcmap_against_cmap <- gess_gcmap(qSig=qsig_gcmap_against_cmap, higher=1, lower=-1, chunk_size=5000)

# Search against LINCS reference database
qsig_gcmap_against_lincs <- qSig(qsig = as.matrix(query), gess_method = "gCMAP", refdb = se_lincs)
gcmap_against_lincs <- gess_gcmap(qSig=qsig_gcmap_against_lincs, higher=1, lower=-1, chunk_size=5000)

# Search against LINCS_sub reference database
qsig_gcmap_against_lincs_sub <- qSig(qsig = as.matrix(query), gess_method = "gCMAP", refdb = se_lincs_sub)
gcmap_against_lincs_sub <- gess_gcmap(qsig_gcmap_against_lincs_sub, higher=1, lower=-1, chunk_size=5000)
```

```{r show_gcmap_res, eval=FALSE}
gcmap_against_cmap <- readRDS("~/insync/project/GESS_and_FEA/results/gcmap_against_cmap.rds")
gcmap_against_cmap@result
gcmap_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/gcmap_against_lincs.rds")
gcmap_against_lincs@result
gcmap_against_lincs_sub <- readRDS("~/insync/project/GESS_and_FEA/results/gcmap_against_lincs_sub.rds")
gcmap_against_lincs_sub@result
```

## Searching with Fisher method for GESS

Generate "qSig" object for GESS_fisher method and search against CMAP, LINCS, LINCS_sub reference database
```{r gess_fisher, eval=FALSE}
query <- as.matrix(assay(se_lincs)[,"sirolimus__MCF7__trt_cp"])

# Search against CMAP reference database
qsig_fisher_against_cmap <- qSig(qsig = as.matrix(query), gess_method = "Fisher", refdb = se_cmap)
fisher_against_cmap <- gess_fisher(qSig=qsig_fisher_against_cmap, higher=1, lower=-1, chunk_size=5000)

# Search against LINCS reference database
qsig_fisher_against_lincs <- qSig(qsig = as.matrix(query), gess_method = "Fisher", refdb = se_lincs)
fisher_against_lincs <- gess_fisher(qSig=qsig_fisher_against_lincs, higher=1, lower=-1, chunk_size=5000)

# Search against LINCS_sub reference database
qsig_fisher_against_lincs_sub <- qSig(qsig = as.matrix(query), gess_method = "Fisher", refdb = se_lincs_sub)
fisher_against_lincs_sub <- gess_fisher(qSig=qsig_fisher_against_lincs_sub, higher=1, lower=-1, chunk_size=5000)
```

```{r show_fisher_res, eval=FALSE}
fisher_against_cmap <- readRDS("~/insync/project/GESS_and_FEA/results/fisher_against_cmap.rds")
fisher_against_cmap@result
fisher_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/fisher_against_lincs.rds")
fisher_against_lincs@result
fisher_against_lincs_sub <- readRDS("~/insync/project/GESS_and_FEA/results/fisher_against_lincs_sub.rds")
fisher_against_lincs_sub@result
```

## Searching with Spearman method for GESS

Generate "qSig" object for GESS_SP method and search against CMAP, LINCS, LINCS_sub reference database

Genome-wide Spearman correlation
```{r gess_sp, eval=FALSE}
# Use genome-wide expression data of sirolimus treatment in MCF7 cells in CMAP_expr database as query
query <- as.matrix(assay(se_cmap_expr)[,"sirolimus__MCF7__trt_cp"])

# Search against CMAP_expr reference database
qsig_sp_against_cmap <- qSig(qsig = as.matrix(query), gess_method = "SP", refdb = se_cmap_expr)
sp_against_cmap <- gess_cor(qSig=qsig_sp_against_cmap, method="spearman", chunk_size=5000)

# Search against LINCS_expr reference database
qsig_sp_against_lincs <- qSig(qsig = as.matrix(query), gess_method = "SP", refdb = se_lincs_expr)
sp_against_lincs <- gess_cor(qSig=qsig_sp_against_lincs, method="spearman", chunk_size=5000)

# Search against LINCS_sub_expr reference database
qsig_sp_against_lincs_sub <- qSig(qsig = as.matrix(query), gess_method = "SP", refdb = se_lincs_sub_expr)
sp_against_lincs_sub <- gess_cor(qSig=qsig_sp_against_lincs_sub, method="spearman", chunk_size=5000)
```

```{r show_sp_res, eval=FALSE}
sp_against_cmap <- readRDS("~/insync/project/GESS_and_FEA/results/sp_against_cmap.rds")
sp_against_cmap@result
sp_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/sp_against_lincs.rds")
sp_against_lincs@result
sp_against_lincs_sub <- readRDS("~/insync/project/GESS_and_FEA/results/sp_against_lincs_sub.rds")
sp_against_lincs_sub@result
```

Spearman sub correlatioin
```{r gess_spsub, eval=FALSE}
# Subset expression values of 150 up and down gene sets from "sirolimus__MCF7__trt_cp" signature.

## get 150 up and down gene sets from "sirolimus__MCF7__trt_cp" signature.
sig <- sort(as.matrix(assay(se_lincs)[,"sirolimus__MCF7__trt_cp"]), decreasing = TRUE)
upset <- names(head(sig, 150)); downset <- names(tail(sig,150))
## get expression values of genes in up and down set
query <- as.matrix(assay(se_cmap_expr)[,"sirolimus__MCF7__trt_cp"])
query_sub <- query[c(upset, downset)]

# Search against CMAP_expr reference database
qsig_spsub_against_cmap <- qSig(qsig = as.matrix(query_sub), gess_method = "SP", refdb = se_cmap_expr)
spsub_against_cmap <- gess_cor(qSig=qsig_spsub_against_cmap, method="spearman", chunk_size=5000)

# Search against LINCS_expr reference database
qsig_spsub_against_lincs <- qSig(qsig = as.matrix(query_sub), gess_method = "SP", refdb = se_lincs_expr)
spsub_against_lincs <- gess_cor(qSig=qsig_spsub_against_lincs, method="spearman", chunk_size=5000)

# Search against LINCS_sub_expr reference database
qsig_spsub_against_lincs_sub <- qSig(qsig = as.matrix(query_sub), gess_method = "SP", refdb = se_lincs_sub_expr)
spsub_against_lincs_sub <- gess_cor(qSig=qsig_spsub_against_lincs_sub, method="spearman", chunk_size=5000)
```

```{r show_spsub_res, eval=FALSE}
spsub_against_cmap <- readRDS("~/insync/project/GESS_and_FEA/results/spsub_against_cmap.rds")
spsub_against_cmap@result
spsub_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/spsub_against_lincs.rds")
spsub_against_lincs@result
spsub_against_lincs_sub <- readRDS("~/insync/project/GESS_and_FEA/results/spsub_against_lincs_sub.rds")
spsub_against_lincs_sub@result
```

# Methods for FEA

## dup_hyperG method for TSEA

Subset top 100 ranking drugs in GESS result, get their target set (with duplication) as query for dup_hypergeometric test for TSEA.
Here choose `lincs_res_vs_lincs` as GESS result.
```{r tsea_dup_hyperG, eval=FALSE}
lincs_against_lincs <- readRDS("~/insync/project/GESS_and_FEA/results/lincs_against_lincs.rds")
drugs <- unique(result(lincs_against_lincs)$pert[1:100])
# GO annotation system
dup_hyperG_res <- tsea_dup_hyperG(drugs = drugs, universe = "Default", type = "GO", ont="MF", pvalueCutoff=0.1, pAdjustMethod="BH", qvalueCutoff = 0.2, minGSSize = 10, maxGSSize = 500)
# KEGG annotation system
dup_hyperG_k_res <- tsea_dup_hyperG(drugs = drugs, universe = "Default", type = "KEGG", pvalueCutoff=0.1, pAdjustMethod="BH", qvalueCutoff = 0.2, minGSSize = 10, maxGSSize = 500)
```

## m_GSEA method for TSEA

Subset top 100 ranking drugs in GESS result as query, m_GSEA method internally get their target set and ranked target list with scores as query for TSEA. 
The scores represent weight of targets in the target set. Here choose `lincs_res_vs_lincs` as GESS result.
```{r tsea_mgsea, eval=FALSE}
# GO annotation system
geneSets <- readRDS("~/insync/project/GESS_and_FEA/data/geneSets_470.rds")
geneList <- readRDS("~/insync/project/GESS_and_FEA/data/geneList.rds")
tmp_res <- fgsea2(pathways=geneSets, stats=geneList, nperm=1000, minSize=5, maxSize=500, gseaParam=1, nproc = 1)

mgsea_res <- tsea_mGSEA(drugs=drugs, type="GO", ont="MF", exponent=1, nPerm=1000, pAdjustMethod="BH", pvalueCutoff=0.9, minGSSize=5, maxGSSize=500)
# KEGG annotation system
mgsea_k_res <- tsea_mGSEA(drugs=drugs, type="KEGG", exponent=1, nPerm=1000, pAdjustMethod="BH", pvalueCutoff=0.5, minGSSize=5, maxGSSize=500)
```

## mabs method for TSEA

Subset top 100 ranking drugs in GESS result as query, mabs method internally get their target set and ranked target list with scores as query for TSEA. 
The scores represent weight of targets in the target set. Here choose `lincs_res_vs_lincs` as GESS result.
```{r tsea_mabs, eval=FALSE}
# GO annotation system
mabs_res <- tsea_mabs(drugs=drugs, type="GO", ont="MF", nPerm=1000, pAdjustMethod="BH", pvalueCutoff=0.05, minGSSize=5, maxGSSize=500)
# KEGG annotation system
mabs_k_res <- tsea_mabs(drugs=drugs, type="KEGG", nPerm=1000, pAdjustMethod="BH", pvalueCutoff=0.5, minGSSize=5, maxGSSize=500)
```

```{r show_tsea_res, eval=FALSE}
dup_hyperG_res <- readRDS("~/insync/project/GESS_and_FEA/results/lg_pt_ego_mf.rds")
dup_hyperG_res@result
dup_hyperG_k_res <- readRDS("~/insync/project/GESS_and_FEA/results/dup_hyperG_k_res.rds")
dup_hyperG_k_res@result

mgsea_res <- readRDS("~/insync/project/GESS_and_FEA/results/mgsea_res.rds")
mgsea_res@result
mgsea_k_res <- readRDS("~/insync/project/GESS_and_FEA/results/mgsea_k_res.rds")
mgsea_k_res@result

mabs_res <- readRDS("~/insync/project/GESS_and_FEA/results/mabs_res.rds")
mabs_res@result
mabs_k_res <- readRDS("~/insync/project/GESS_and_FEA/results/mabs_k_res.rds")
mabs_k_res@result
```

## hyperG method for DSEA

Subset top 100 ranking drugs in GESS result as query for hypergeometric test for DSEA, here choose `lincs_res_vs_lincs` as GESS result.
```{r dsea_hyperG, eval=FALSE}
drugs <- unique(result(lincs_against_lincs)$pert[1:100])
# GO annotation system
hyperG_res <- dsea_hyperG(drugs = drugs, type = "GO", ont="MF", pvalueCutoff=0.1, pAdjustMethod="BH", qvalueCutoff = 0.2, minGSSize = 10, maxGSSize = 500)
# KEGG annotation system
hyperG_k_res <- dsea_hyperG(drugs = drugs, type = "KEGG", pvalueCutoff=0.1, pAdjustMethod="BH", qvalueCutoff = 0.2, minGSSize = 10, maxGSSize = 500)
```

## GSEA method for DSEA

Use ranked drug list in GESS result as query for GSEA for DSEA, the scores are similarity scores of corresponding GESS methods. Zeros are removed. Here choose `lincs_res_vs_lincs` as GESS result.
```{r dsea_gsea, eval=FALSE}
dl <- abs(lincs_against_lincs@result$NCS); names(dl) <- lincs_against_lincs@result$Pert
dl <- dl[dl>0]
dl <- dl[!duplicated(names(dl))]
# GO annotation system
gsea_res <- dsea_GSEA(drugList=dl, type="GO", ont="MF", exponent=1, nPerm=1000, pAdjustMethod="BH", pvalueCutoff=0.1, minGSSize=5, maxGSSize=500)

# KEGG annotation system
gsea_k_res <- dsea_GSEA(drugList=dl, type="KEGG", exponent=1, nPerm=1000, pAdjustMethod="BH", pvalueCutoff=0.1, minGSSize=5, maxGSSize=500)
```


```{r show_dsea_res, eval=FALSE}
hyperG_res <- readRDS("~/insync/project/GESS_and_FEA/results/hyperG_res.rds")
hyperG_res@result
hyperG_k_res <- readRDS("~/insync/project/GESS_and_FEA/results/hyperG_k_res.rds")
hyperG_k_res@result

gsea_res <- readRDS("~/insync/project/GESS_and_FEA/results/gsea_res.rds")
gsea_res@result
gsea_k_res <- readRDS("~/insync/project/GESS_and_FEA/results/gsea_k_res.rds")
gsea_k_res@result
```

# Visulization

## Construct drug-target interaction networks in interested GO categories

Build drug-target networks in top ranking GO categories in `hyperG_res`
```{r dtnet_go, eval=FALSE}
dtnetplot(drugs = hyperG_res@drug, set = "GO:0030169", ont = "MF")
```

## Construct drug-target interaction networks in interested KEGG pathways or defined other gene/protein sets

Build drug-target networks in top ranking KEGG pathways in `hyperG_k_res`
```{r dtnet_kegg, eval=FALSE}
dtnetplot(drugs = hyperG_k_res@drug, set = "hsa04979")
```
