#' Build CMAP, LINCS or custome reference database
#' 
#' @title build_db
#' @param df data.frame, represents genoime-wide GEPs (log2FC, Z-scores, intensity values, etc.) of compound or genetic treatments. 
#' Rownames (Gene IDs) should be included. The column names should be of `(drug)__(cell)__(factor)` format, e.g., `sirolimus__MCF7__trt_cp`.
#' @param file a character string, path to the tabular file in tab-delimited format or gctx file for LINCS database. The tabular file 
#' can mimic both count data from RNA-Seq and intensity values from gene chips, Rownames (Gene IDs) should be included. 
#' The column names should be of `(drug)__(cell)__(type)` format. The tabular file 
#' for CMAP differential expression (DE) database (`logMA.tsv`) can be downloaded at \url{http://biocluster.ucr.edu/~yduan004/CMAP_db/logMA.tsv}. 
#' The `logMA` matrix contains the log2FC scores generated from DE analysis with `limma` 
#' from the original CEL files generated by Affymetrix Chips. The CMAP Affy data is stored at \url{https://portals.broadinstitute.org/cmap/}.
#' The code used for DE analysis is vailable at \url{http://girke.bioinformatics.ucr.edu/longevityTools/mydoc/mydoc_longevityTools_eDRUG_05.html}.
#' 
#' The tabular file for CMAP expression database (`CMAP_expr.tsv`) can be downloaded at \url{http://biocluster.ucr.edu/~yduan004/CMAP_db/CMAP_expr.tsv}. 
#' The `CMAP_expr` matrix contains the rma normalized gene expression intensity values for drug treatments in different cells. To combine 
#' expression values of a drug treatment at diffenrent concentration and duplications in cell, It takes the mean values.
#' The code used to generate `CMAP_expr.tsv` is vailable at \url{}.
#' 
#' The gctx files for LINCS database can be downloaded at \url{https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742}. The Level3 file 
#' (GSE92742_Broad_LINCS_Level3_INF_mlr12k_n1319138x12328.gctx) contains gene expression value of treated and untreated samples. 
#' Here we take the mean expression values of compound treament samples at 10 uM, 24h as signature of that compound treatment in cells.
#' The Level5 file (GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx) contains moderate z-scores from compound or genetic treatments 
#' from DE analysis. The compound treatment are at different concentation, time and also includes duplications. Here, we only subset thoese 
#' compound treatments that are at 10 uM, 24h, and eliminate the technical duplications.
#' Note, the gctx files for LINCS database is very large (~25 Gb), it is recommended to use biocluster and request enough memory to run the analysis   
#' 
#' @param annot data.frame or path to the tabular file containing sample annotation information, which means column annotation for database 
#' in `file` parameter. The row names of the data.frame in `annot` should match the column names of data.frame in `file`. For LINCS database, 
#' if `file` is Level3, the `annot` should be path to the "GSE92742_Broad_LINCS_inst_info.txt" file. 
#' If `file` is Level5, the `annot` should be path to the "GSE92742_Broad_LINCS_sig_info.txt" file. In addition, the cell type information will be
#' automatically added.
#' @param type one of `CMAP`, `LINCS` or `Custom`. If `type` is `CMAP`, `file` should be the path to the downloaded `logMA.tsv`. If `type` is `LINCS`, 
#' `file` should be the path to the downloaded Level3 or Level5 gctx file at GSE92742
#' @param dest_path path to store the reference database, which is the hdf5 backed \code{SummarizedExperiment} object
#' @return hdf5 backed \code{SummarizedExperiment} object
#' @export

build_db <- function(df=NULL, file=NULL, annot=NULL, type, dest_path){
  if (type == "CMAP"){
    df <- read.delim(file = file, sep = "\t", row.names = 1, check.names = FALSE)
    pert_cell_factor = gsub("(^.*)_(.*$)", "\\1__\\2__trt_cp", colnames(df))
    colnames(df)=pert_cell_factor
    cmap <- SummarizedExperiment(assays = list(score=as.matrix(df)))
    ## Save 'cmap' as an HDF5-based SummarizedExperiment object:
    h5_cmap <- saveHDF5SummarizedExperiment(cmap, dest_path)
    return(h5_cmap)
  }
  if (type == "LINCS"){
    if (grepl("Level5", file)){
      
      
    }
    if (grepl("Level3", file)){
      
      
    }
  }
  if(type == "Custom"){
    df <- read.delim(file = file, sep = "\t", row.names = 1, check.names = FALSE)
    if(is.character(annot)){
      annot <- read.delim(file = annot, sep = "\t", row.names = 1, check.names = FALSE)
    }
    col_anno <- as(annot, "DataFrame")
    se <- SummarizedExperiment(assays = list(score=as.matrix(df)), colData = col_anno)
    ## Save 'cmap' as an HDF5-based SummarizedExperiment object:
    h5_se <- saveHDF5SummarizedExperiment(se, dest_path)
    return(h5_se)
  }
}